lib/streamlit/server/Server.py

```diff
@@ -86,6 +89,44 @@ class State(Enum):
     STOPPED = "STOPPED"
 
 
+class ExceededRetries(Exception):
+    pass
+
+
+def server_port_is_manually_set():
+    return config.is_manually_set("server.port")
+
+
+def start_listening(app, call_count=0):
+    """makes the server start listening at the configured port.
+    In case the port is already taken it tries listening to the next available port.
+    It will error after a 100 attempts."""
+    port = config.get_option("server.port")
+    try:
+        app.listen(port)
+    except OSError as e:
+        if e.errno == errno.EADDRINUSE:
+            if call_count >= 100:
+                raise ExceededRetries(
+                    "Port %s already in use, were unable to find a free port after a hundred attempts",
+                    port,
+                )
+            if server_port_is_manually_set():
+                LOGGER.debug("Port %s already in use, trying to use the next one", port)
```

Comment:
When the server port is manually set we shouldn't try other ports. Just show an error and let Streamlit quit.

Edit: actually I think you got the branches of this if/else mixed up.

Reply:
fixed